# async

async函数是Generator函数的语法糖。 

    async function timeout(ms) {
        await new Promise((resolve) => {
            setTimeout(resolve, ms); 
        }); 
    }
    async function asyncPrint(value, ms) {
        await timeout(ms); 
        console.log(value); 
    }
    asyncPrint('hello world', 50); 

async函数的返回值都是 `Promise` 对象(不是也会被转成立刻 `resolve` 的 `Promise` ), 可以用then方法指定下一步的操作。 

    async function f() {
        return await 123; 
    }
    f().then(v => console.log(v)) // 123

此外, 除了当 `await` 函数返回了 `reject` 会阻止后续代码的执行, 想要继续后面的代码, 只能使用 `try...catch` 或者直接在Promise后加 `rejcte` 。 

    async function f() {
        await Promise.reject('error happend').catch(e => console.log(e)); 
        return await Promise.resolve('hello world'); 
    }
    async function f_try() {
        await Promise.reject('error happend').catch(e => console.log(e)); 
        return await Promise.resolve('hello world'); 
    }
    f().then(v => console.log(v))
    // error happend
    // hello world

关于多

    // 多任务并行
    let [foo, bar] = await Promise.all([getFoo(), getBar()]); 
    // 写法二
    let fooPromise = getFoo(); 
    let barPromise = getBar(); 
    let foo = await fooPromise; 
    let bar = await barPromise; 

